traffic_ops/traffic_ops_golang/trafficstats/trafficstats.go

```diff
@@ -0,0 +1,691 @@
+	WHERE cachegroup = 'total'
+	AND deliveryservice = $xmlid
+	AND time >= $start
+	AND time <= $end
+	GROUP BY time(%s, %s), cachegroup%s`
+
+type orderable string
+
+const (
+	timeOrder  orderable = "time"
+	valueOrder orderable = "value"
+)
+
+func orderableFromString(v string) *orderable {
+	var o orderable
+	switch v {
+	case "time":
+		o = timeOrder
+	case "value":
+		o = valueOrder
+	default:
+		return nil
+	}
+	return &o
+}
+
+type config struct {
+	DeliveryService string
+	End             time.Time
+	ExcludeSeries   bool
+	ExcludeSummary  bool
+	Interval        tc.TrafficStatsDuration
+	Limit           *uint64
+	MetricType      string
+	Offset          *uint64
+	OrderBy         *orderable
+	Start           time.Time
+	Unix            bool
+}
+
+func configFromRequest(r *http.Request, i *api.APIInfo) (c config, e error, code int) {
+	code = http.StatusBadRequest // this is so common I may as well do it right away
+
+	if accept := r.Header.Get("Accept"); accept != "" {
+
+		mimes, err := rfc.MimeTypesFromAccept(accept)
+		if err != nil {
+			log.Warnf("Failed to negotiate content, Accept line '%s', error: %v", accept, err)
+		} else {
+
+			found := false
+			for _, m := range mimes {
+				if jsonWithRFCTimestamps.Equal(m) {
+					found = true
+					break
+				}
+
+				if jsonWithUnixTimestamps.Equal(m) {
+					found = true
+					c.Unix = true
+					break
+				}
+			}
+
+			if !found {
+				e = fmt.Errorf("Failed to negotiate content; cannot produce output satisfying %s", accept)
+				code = http.StatusNotAcceptable
+				return
+			}
+		}
+	}
+
+	if limit, ok := i.Params["limit"]; ok {
+		lim, err := strconv.ParseUint(limit, 10, 64)
+		if err != nil {
+			e = errors.New("Invalid limit!")
+			return
+		}
+		c.Limit = &lim
+	}
+
+	if offset, ok := i.Params["offset"]; ok {
+		off, err := strconv.ParseUint(offset, 10, 64)
+		if err != nil {
+			e = errors.New("Invalid offset!")
+			return
+		}
+		c.Offset = &off
+	}
+
+	if orderby, ok := i.Params["orderby"]; ok {
+		if c.OrderBy = orderableFromString(orderby); c.OrderBy == nil {
+			e = errors.New("Invalid orderby!")
+			return
+		}
+	}
+
+	if c.Start, e = parseTime(i.Params["startDate"]); e != nil {
+		log.Errorf("Parsing startDate: %v", e)
+		e = errors.New("Invalid startDate!")
+		return
+	}
+
+	if c.End, e = parseTime(i.Params["endDate"]); e != nil {
+		log.Errorf("Parsing endDate: %v", e)
+		e = errors.New("Invalid endDate!")
+		return
+	}
+
+	if interval, ok := i.Params["interval"]; !ok {
+		c.Interval = DEFAULT_INTERVAL
+	} else if c.Interval = tc.TrafficStatsDurationFromString(interval); c.Interval == tc.InvalidDuration {
+		log.Errorf("Error parsing 'interval' query parameter: %v", e)
+		e = errors.New("Invalid interval!")
+		return
+	}
+
+	if ex, ok := i.Params["exclude"]; ok {
+		switch ex {
+		case "series":
+			c.ExcludeSeries = true
+		case "summary":
+			c.ExcludeSummary = true
+		default:
+			e = errors.New("Invalid exclude! Must be 'series' or 'summary'")
```

Comment:
should this have been a function like orderableFromString? one used switch in place and returns options to user and the other is func call and doesnt return options to user

Reply:
Yes, that should be consistent, and they should both return options.

Reply:
sure! 👍